### PR TITLE
feat(470): add template model

### PR DESCRIPTION
**Pull request by minzcmu**
~~Blocked by changes to be done in data-schema~~

Related to https://github.com/screwdriver-cd/screwdriver/issues/470

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -700,6 +700,124 @@ Get builds that belong to this event
...
+const factory = Model.TemplateFactory.getInstance({
+    datastore
+});
+const config = {
+    name: 'testTemplate',
+    version: '1.3',
+    description: 'I am a test template',    
+    maintainer: 'foo@bar.com',
+    scmUri: 'github:123:master',
+    config: 'image: node:6',
```
Comment (d2lam) [inline README.md]: object

File: README.md [hunk 2]
```diff
@@ -700,6 +700,124 @@ Get builds that belong to this event
...
+```
+
+| Parameter        | Type  |  Required | Description |
+| :-------------   | :---- | :-------------|  :-------------|
+| config        | Object | Yes | Configuration Object |
+| config.name | String | Yes | The template name |
+| config.version | String | Yes | Version of the template |
+| config.description | String | Yes | Description of the template |
+| config.maintainer | Array | Yes | Maintainer's email |
+| config.config | String | Yes | Config of the screwdriver-template.yaml |
```
Comment (d2lam) [inline README.md]: object

File: lib/templateFactory.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * Instantiate a Template class
+     * @method createClass
+     * @param  {Object}     config               Template data
+     * @param  {Datastore}  config.datastore     Object that will perform operations on the datastore
+     * @param  {String}     config.name          The template name
+     * @param  {String}     config.version       Version of the template
+     * @param  {String}     config.description   Description of the template
+     * @param  {String}     config.maintainer    Maintainer's email
+     * @param  {String}     config.config        Config of the screwdriver-template.yaml
```
Comment (d2lam) [inline lib/templateFactory.js]: object

File: lib/templateFactory.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+     * @param  {String}     config.scmUri        ScmUri of the template pipeline
+     * @param  {Array}      config.labels        Labels attached to the template
+     * @return {Promise}
+     */
+    create(config) {
+        const c = config;
```
Comment (d2lam) [inline lib/templateFactory.js]: remove this

**Review by d2lam — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -700,6 +700,64 @@ Get builds that belong to this event
...
 
+### Template Model
+```js
+'use strict';
+const Model = require('screwdriver-models');
+const factory = Model.TemplateFactory.getInstance({
+    datastore
+});
+const config = {
+    name: 'testTemplate',
```
Comment (FenrirUnbound) [inline README.md:711]: Question: does this mean that we're not namespacing templates?

Comment (FenrirUnbound) [reply README.md:711]: I would have expected a `namespace/template_name` format being enforced

File: lib/templateFactory.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+     * @method createClass
+     * @param  {Object}     config               Template data
+     * @param  {Datastore}  config.datastore     Object that will perform operations on the datastore
+     * @param  {String}     config.name          The template name
+     * @param  {String}     config.version       Version of the template
+     * @param  {String}     config.description   Description of the template
+     * @param  {String}     config.maintainer    Maintainer's email
+     * @param  {Object}     config.config        Config of the screwdriver-template.yaml
+     * @param  {String}     config.scmUri        ScmUri of the template pipeline
+     * @param  {Array}      config.labels        Labels attached to the template
```
Comment (FenrirUnbound) [inline lib/templateFactory.js]: The `README` states that this is optional, but the JSDoc does not reflect that. Which one is it?/

**Review by FenrirUnbound — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (tkyi) [reply README.md:711]: I'm not sure if we ever figured out how to police/enforce naming conventions for templates

Comment (tkyi) [reply README.md:711]: But yeah I think St. John was envisioning some sort of `platform/step` sort of format

File: lib/templateFactory.js (hunk 6 above)
Comment (minzcmu) [reply lib/templateFactory.js]: fixed. Should be optional

File: README.md (hunk 5 above)
Comment (minzcmu) [reply README.md:711]: @tkyi @FenrirUnbound I am not sure if we should enforce platform/step or let the platform owner to decide what is the proper format. If `platform/step` then we need to change the `data-schema`

Comment (minzcmu) [reply README.md:711]: The naming issue will be addressed in next pass.

**Review by FenrirUnbound — APPROVED** [review]
👍 

giphy 51

**Review by d2lam — APPROVED** [review]
LGTM 👍

**Pull request merged**